#!/usr/bin/env bash

# Collapse super dups to within 150kbp and filter for those <= 150kbp in length from start to end:
echo "Collapsing segmental duplications..."
sort -k1,1 -k2,2n genomicSuperDups.txt | bedtools merge -d 150000 -i - | awk '($3-$2)<=150000' | grep -v "random" | grep -v "Un" > genomicSuperDups.sorted.merged.150kbp.txt

# obtain SVs in SDs:
echo ""
echo ""
echo "Obtaining filtered OM SVs crossing within segmental duplications..."
bedtools intersect -wa -wb -a optical_mapping_filtered_svs.txt -b genomicSuperDups.sorted.merged.150kbp.txt > optical_mapping_filtered_svs_in_sds.txt

# obtain breakpoints regarding the sizes of segmental duplications and SVs; 5' most and 3' most loci are chosen:
awk '{if ($2>=$6 && $3<=$7) {print $4"\t"$5"\t"$6"\t"$7} else if ($2>=$6 && $3>=$7) {print $4"\t"$5"\t"$6"\t"$3} else if ($2<=$6 && $3<=$7) {print $4"\t"$5"\t"$2"\t"$7} else if ($2<=$6 && $3>=$7) {print $4"\t"$5"\t"$2"\t"$3}}' optical_mapping_filtered_svs_in_sds.txt | sed 's/chr//g' | sed 's/X/23/g' | sed 's/Y/24/g' > optical_mapping_filtered_svs_in_sds_scenarios.txt

# STARTS:
while read line1 line2 line3 line4; do c=$(echo $line2 | sed 's/chr//g'); awk -v info="$line1" -v chr="$c" -v start="$line3" -v end="$line4" '$1 == chr {printf "%s\t%d\t%d\t%d\t%.1f\t%.1f\t%d\n", info,chr,start,end,start-$6,$6,$4}' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1_r.cmap | sort -k5n | grep -v "-" | head -5 | tr ',' '\t' | awk -v i="$line1" '{print $7","$8","$9","i","$2","$12}' | awk 'NR%5{printf "%s\t",$0;next;}1'; done <optical_mapping_filtered_svs_in_sds_scenarios.txt > optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels.txt

while read line1 line2 line3 line4 line5; do og=$(echo $line1 | cut -d',' -f1-3); info=$(echo $line1 | cut -d',' -f4-9); contig=$(echo $line1 | cut -d',' -f10); label1=$(echo $line1 | cut -d',' -f11); label2=$(echo $line2 | cut -d',' -f11); label3=$(echo $line3 | cut -d',' -f11); label4=$(echo $line4 | cut -d',' -f11); label5=$(echo $line5 | cut -d',' -f11); if awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label1,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label1,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label2,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label2,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label3,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label3,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label4,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label4,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label5,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label5,[0-9]+" | head -1 | cut -d',' -f2; fi; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels.txt | awk 'NR%3{printf "%s\t",$0;next;}1' > optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels_contigid_contiglabel.txt

# ENDS: WORKS
while read line1 line2 line3; do c=$(echo $line1 | cut -d',' -f1); s=$(echo $line1 | cut -d',' -f2); e=$(echo $line1 | cut -d',' -f3); awk -v info="$line2" -v chr="$c" -v start="$s" -v end="$e" '$1 == chr {printf "%s\t%d\t%d\t%d\t%.1f\t%.1f\t%d\n", info,chr,start,end,$6-end,$6,$4}' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1_r.cmap | sort -k5n | grep -v "-" | head -5 | tr ',' '\t' | awk -v i="$line2" '{print $7","$8","$9","i","$2","$12}' | awk 'NR%5{printf "%s\t",$0;next;}1'; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels_contigid_contiglabel.txt > optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels.txt

while read line1 line2 line3 line4 line5; do og=$(echo $line1 | cut -d',' -f1-3); info=$(echo $line1 | cut -d',' -f4-9); contig=$(echo $line1 | cut -d',' -f10); label1=$(echo $line1 | cut -d',' -f11); label2=$(echo $line2 | cut -d',' -f11); label3=$(echo $line3 | cut -d',' -f11); label4=$(echo $line4 | cut -d',' -f11); label5=$(echo $line5 | cut -d',' -f11); if awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label1,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label1,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label2,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label2,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label3,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label3,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label4,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label4,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label5,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label5,[0-9]+" | head -1 | cut -d',' -f2; fi; done <optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels.txt | awk 'NR%3{printf "%s\t",$0;next;}1' > optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels_contigid_contiglabel.txt

# Go back and apply new filters for the starts:
while read line1 line2 line3; do c=$(echo $line1 | cut -d',' -f1); s=$(echo $line1 | cut -d',' -f2); e=$(echo $line1 | cut -d',' -f3); awk -v info="$line2" -v chr="$c" -v start="$s" -v end="$e" '$1 == chr {printf "%s\t%d\t%d\t%d\t%.1f\t%.1f\t%d\n", info,chr,start,end,start-$6,$6,$4}' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1_r.cmap | sort -k5n | grep -v "-" | head -5 | tr ',' '\t' | awk -v i="$line2" '{print $7","$8","$9","i","$2","$12}' | awk 'NR%5{printf "%s\t",$0;next;}1'; done <optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels_contigid_contiglabel.txt > optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels.txt

while read line1 line2 line3 line4 line5; do og=$(echo $line1 | cut -d',' -f1-3); info=$(echo $line1 | cut -d',' -f4-9); contig=$(echo $line1 | cut -d',' -f10); label1=$(echo $line1 | cut -d',' -f11); label2=$(echo $line2 | cut -d',' -f11); label3=$(echo $line3 | cut -d',' -f11); label4=$(echo $line4 | cut -d',' -f11); label5=$(echo $line5 | cut -d',' -f11); if awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label1,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label1,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label2,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label2,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label3,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label3,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label4,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label4,[0-9]+" | head -1 | cut -d',' -f2; elif awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -q "($label5,"; then echo $og; echo $info; awk -v con="$contig" '$2 == con' ../../exp_refineFinal1/alignref_final/EXP_REFINEFINAL1.xmap | grep -Eo "$label5,[0-9]+" | head -1 | cut -d',' -f2; fi; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels.txt | awk 'NR%3{printf "%s\t",$0;next;}1' > optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels_contigid_contiglabel.txt

# Obtain the contig positions for the contig labels found in the above 2 files:
while read line1 line2 line3; do id=$(echo $line2 | cut -d',' -f2); awk -v label="$line3" -v info="$line2" -v og="$line1" '$4 == label {print og"\t"label"\t"$6}' ../../../exp_refineFinal1_contig"$id"_r.cmap | head -1; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels_contigid_contiglabel.txt > optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels_contigpos.txt
while read line1 line2 line3; do id=$(echo $line2 | cut -d',' -f2); awk -v label="$line3" -v info="$line2" '$4 == label {print label"\t"$6"\t"info}' ../../../exp_refineFinal1_contig"$id"_r.cmap | head -1; done <optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels_contigid_contiglabel.txt > optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels_contigpos.txt

paste optical_mapping_filtered_svs_in_sds_scenarios_starts_5labels_contigpos.txt optical_mapping_filtered_svs_in_sds_scenarios_ends_5labels_contigpos.txt > optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos.txt

# NOTE: some of the starts will be after the ends due to orientation of contig alignment to reference (- vs +):
awk '{if ($5 > $3) {print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6} else {print $1"\t"$2"\t"$5"\t"$4"\t"$3"\t"$6}}' optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos.txt | awk 'NF=="6"' > optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos_fixed.txt

# Counts number of molecules crossing an SV:
while read line1 line2 line3 line4 line5 line6; do id=$(echo $line6 | cut -d',' -f2); start="$line3"; end="$line5"; awk -v i="$id" -v s="$start" -v e="$end" '{ if(s >= $6 && s <= $7) {print $0} else if(e >= $6 && e <= $7) {print $0} }' ../../../exp_refineFinal1_contig"$id".xmap | wc -l | sed 's/\ //g'; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos_fixed.txt > optical_mapping_filtered_svs_in_sds_number_molecules.txt

# Confidence of molecules crossing into the SV:
while read line1 line2 line3 line4 line5 line6; do id=$(echo $line6 | cut -d',' -f2); start="$line3"; end="$line5"; awk -v i="$id" -v s="$start" -v e="$end" '{ if(s >= $6 && s <= $7) {print $9} else if(e >= $6 && e <= $7) {print $9} }' ../../../exp_refineFinal1_contig"$id".xmap | sed 's/\ //g' | tr '\n' ',' | sed 's/,$//g'; echo ""; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos_fixed.txt > optical_mapping_filtered_svs_in_sds_confidence_only.txt

# Counts 5’ labels flanking an SV:
#cut -f4 optical_mapping_filtered_svs_not_in_sds.txt | cut -d',' -f2,3,4 | tr ',' '\t' > tmp_contig_ids
#while read line1 line2 line3 line4 line5; do awk -v i="$line1" -v s="$line2" -v e="$line3" 'function abs(x){return ((x < 0.0) ? -x : x)} {if ($1==i && abs($6-s) < 1) {print i","s","e","$4} else if($1==i && abs($6-e) < 1) {print i","s","e","$4}}' exp_refineFinal1_merged_q.cmap | tr '\n' '\t'; echo ""; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos_fixed.txt > tmp_contig_ids_labels

# Before labels 1 and 2 and after labels 1 and 2:
while read line1 line2 line3 line4 line5 line6; do id=$(echo $line6 | cut -d',' -f2); label1="$line2"; label2="$line4"; grep -v "#" ../../../exp_refineFinal1_contig"$id".xmap | cut -f14 | grep "($label1," | while read line; do echo $line | sed 's/(//g' | tr ')' '\n' | awk 'NF > 0' | cut -d',' -f1 | sed -n "0,/$label1/p" | wc -l; done < "${1:-/dev/stdin}" | awk '{$1=$1};1' | tr '\n' ',' | sed 's/,$//g'; echo "" | tr '\n' '\t'; grep -v "#" ../../../exp_refineFinal1_contig"$id".xmap | cut -f14 | grep "($label2," | while read line; do echo $line | sed 's/(//g' | tr ')' '\n' | awk 'NF > 0' | cut -d',' -f1 | sed -n "0,/$label2/p" | wc -l; done < "${1:-/dev/stdin}" | awk '{$1=$1};1' | tr '\n' ',' | sed 's/,$//g'; echo "" | tr '\n' '\t'; grep -v "#" ../../../exp_refineFinal1_contig"$id".xmap | cut -f14 | grep "($label1," | while read line; do echo $line | sed 's/(//g' | tr ')' '\n' | awk 'NF > 0' | cut -d',' -f1 | sed "1,/$label1/d" | wc -l; done < "${1:-/dev/stdin}" | awk '{$1=$1};1' | tr '\n' ',' | sed 's/,$//g'; echo "" | tr '\n' '\t'; grep -v "#" ../../../exp_refineFinal1_contig"$id".xmap | cut -f14 | grep "($label2," | while read line; do echo $line | sed 's/(//g' | tr ')' '\n' | awk 'NF > 0' | cut -d',' -f1 | sed "1,/$label2/d" | wc -l; done < "${1:-/dev/stdin}" | awk '{$1=$1};1' | tr '\n' ',' | sed 's/,$//g'; echo ""; done <optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos_fixed.txt > svs_in_sds_tmp_contig_ids_labels_before_after_label1_and_label2

# Paste together all files:
paste optical_mapping_filtered_svs_in_sds_scenarios_starts_ends_5labels_contigpos_fixed.txt optical_mapping_filtered_svs_in_sds_number_molecules.txt optical_mapping_filtered_svs_in_sds_confidence_only.txt svs_in_sds_tmp_contig_ids_labels_before_after_label1_and_label2 > svs_in_sds_support.txt

